src/base/static/components/templates/map.tsx

```diff
@@ -173,6 +173,7 @@ type Props = StateProps &
   TransProps;
 
 class MapTemplate extends Component<Props, State> {
+  private mainMapRef = createRef<HTMLDivElement>();
```

Comment:
I'm very wary of passing refs as props, especially drilling through so many of our components - I think it's a slippery slope and will make the code difficult to maintain.

This is definitely a tricky use case - I would have to spend some time to come up with a better pattern :thinking:  

But for now, I think this is an improvement:

Since we only need the function `mainMap.querySourceFeatures`, I think we keep `mainMapRef` inside of MainMap, and add a method on MainMap like this:

```js
componentDidMount() {
   this.map = this.mapRef.current.getMap();
   // ...etc
   this.props.setQuerySourceFeatures(this.map.querySourceFeatures);
  // ..etc
}
```

Then we can add a prop on MainMap like so:

```js
<MainMap
    setQuerySourceFeatures={ querySourceFeatures => this.querySourceFeatures = querySourceFeatures }
    // ...etc
/>
```

which will allow us to pass a prop to ContentPanel like so:

```js
<ContentPanel
  querySourceFeatures={this.querySourceFeatures}
  // ...etc
```

That should avoid leaking the MainMap ref outside of the MainMap component.

Reply:
I was uncertain about the map ref as well. I like this idea of only exposing one method on the map ref-- seems like an improvement.

There's definitely an architectural question looming, though, because I think to drive most or all of the geospatial "intelligence" features which we might want we'll need access to the imperative API of the Mapbox instance, which can only be accessed via a ref. Another use case would be geospatially-aware dashboard widgets.

But maybe we can use the pattern you suggest above to selectively expose select Mapbox API methods outside `MainMap`.